### PR TITLE
Fixed dependency of TransactionalMapProxy on MapContainer

### PR DESCRIPTION
**Pull request by Donnerbart**
My solution to this issue is to move some logic from the `MapContainer` to the `MapServiceContext`. Since the context is shared between all maps, I used a `ConcurrentMap<String, ?>` to cache the migrated state in the context. So it doesn't matter if the `MapContainer` or `TransactionalMapProxy` wants to access the field first. The cached data is destroyed when the map is destroyed, so there should be no leak.

Fixes https://github.com/hazelcast/hazelcast/issues/10254

This test is still failing, which is passing on the original `MapProxyImplTest`:
**TransactionalMapProxyTest.whenMapProxyWithEagerMapStoreIsCreated_mapContainerIsNotCreated()**
```java
java.lang.AssertionError: expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.map.impl.tx.TransactionalMapProxyTest.assertNoMapContainersExist(TransactionalMapProxyTest.java:102)
	at com.hazelcast.map.impl.tx.TransactionalMapProxyTest.whenMapProxyWithEagerMapStoreIsCreated_mapContainerIsNotCreated(TransactionalMapProxyTest.java:68)
```

### CODE REVIEW

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
@Donnerbart I'm not sure we can avoid creating a map container for the scenario in the failing test. The `IsKeyLoadFinishedOperation` creates the record store. It is triggered when waiting for the store to be loaded on proxy creation : https://github.com/hazelcast/hazelcast/blob/6700e09773743d453d18a0219bfcef860e9468a9/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java#L299
To cover this scenario, the `PartitionContainer#createRecordStore` shouldn't instantiate a `MapContainer`

**Comment by Donnerbart** [comment]
It's been awhile, but I think I came to the same conclusion about that test. I had a quick peek at the `PartitionContainer`, some things could be replaced to not depend on the `MapContainer`, but in the end the `RecordStore` needs it (and if we don't want to get rid of it completely, at some point we will anyway need to create it).

So I've ignored the failing test for now:
```java
    @Ignore(value = "PartitionContainer#createRecordStore() creates a MapContainer, which we cannot avoid for now")
    public void whenMapProxyWithEagerMapStoreIsCreated_mapContainerIsNotCreated() {
```

File: hazelcast/src/main/java/com/hazelcast/map/impl/record/RecordComparator.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.record;
+
+public interface RecordComparator {
+
+    boolean isEquals(Object value1, Object value2);
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/map/impl/record/RecordComparator.java]: minor: maybe rename to `isEqual` or `equals`? It was `isEquals` previously in `RecordFactory` but "is equals" sounds a bit weird to me.

**Comment by vbekiaris** [comment]
EE side counterpart?

**Comment by Donnerbart** [comment]
EE counterpart is this one: https://github.com/hazelcast/hazelcast-enterprise/pull/1651

File: hazelcast/src/main/java/com/hazelcast/map/impl/record/RecordComparator.java (hunk 1 above)
Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/map/impl/record/RecordComparator.java]: Hmm, I think `equals()` should be reserved to compare an instance with another instance of itself (together with `hashCode()`). But `isEqual()` sounds indeed better, I've changed it.

**Pull request merged**